internal/stream/stream_test.go

```diff
@@ -250,3 +251,90 @@ func TestPushStream_Stop(t *testing.T) {
 	err := ps.Stop()
 	assert.Nil(t, err)
 }
+
+func TestNewPushStream_Failure(t *testing.T) {
+	ctx := context.Background()
+	ctrl := gomock.NewController(t)
+	subscriptionCoreMock := mocks1.NewMockICore(ctrl)
+	subscriberCoreMock := mocks2.NewMockICore(ctrl)
+	workerID := uuid.New().String()
+	httpConfig := &httpclient.Config{}
+	subscriptionCoreMock.EXPECT().Get(gomock.Any(), subName).Return(nil, errors.New("Test Error"))
+	got, err := NewPushStream(ctx, workerID, subName, subscriptionCoreMock, subscriberCoreMock, httpConfig)
+	assert.Nil(t, got)
+	assert.NotNil(t, err)
+}
+
+func TestStart_RestartSubscriber(t *testing.T) {
+	ctx := context.Background()
+	ctrl := gomock.NewController(t)
+	ps := getMockPushStreamRestart(t, ctx, ctrl, "")
+	sub := ps.subs
+	go ps.Start()
+	<-time.NewTicker(10 * time.Millisecond).C
+	assert.NotNil(t, sub)
+	ps.subs.GetErrorChannel() <- errors.New("Subscriber Test Error")
+	select {
+	case <-time.NewTicker(1 * time.Second).C:
+		assert.NotNil(t, ps.subs)
+		assert.NotEqual(t, sub, ps.subs)
+		ps.Stop()
+	}
+}
+
+func getMockPushStreamRestart(t *testing.T, ctx context.Context, ctrl *gomock.Controller, endpoint string) *PushStream {
+	subscriptionCoreMock := mocks1.NewMockICore(ctrl)
+	subscriberCoreMock := mocks2.NewMockICore(ctrl)
+	subModel := getMockSubModel(endpoint)
+	subscriptionCoreMock.EXPECT().Get(gomock.Any(), subName).Return(subModel, nil)
+	workerID := uuid.New().String()
+	httpConfig := &httpclient.Config{}
+	pushStream, _ := NewPushStream(ctx, workerID, subName, subscriptionCoreMock, subscriberCoreMock, httpConfig)
+	pushStream.subs = getMockSubscriberRestart(t, ctx, ctrl)
+	subscriberCoreMock.EXPECT().NewSubscriber(
+		ctx,
+		workerID,
+		subModel,
+		defaultTimeoutMs,
+		defaultMaxOutstandingMsgs,
+		defaultMaxOuttandingBytes,
+		gomock.AssignableToTypeOf(make(chan *subscriber.PullRequest)),
+		gomock.AssignableToTypeOf(make(chan *subscriber.AckMessage)),
+		gomock.AssignableToTypeOf(make(chan *subscriber.ModAckMessage))).Return(getMockSubscriberRestart(t, ctx, ctrl), nil).Times(2)
+	return pushStream
+}
+
+func getMockSubscriberRestart(t *testing.T, ctx context.Context, ctrl *gomock.Controller) *mocks3.MockISubscriber {
+	subscriberMock := mocks3.NewMockISubscriber(gomock.NewController(t))
```

Comment:
do add a comment on why we need to initialize with a new controller here

Reply:
Added